File: app/src/main/java/ru/mrlargha/stemobile/ui/users/UsersViewModel.java
package ru.mrlargha.stemobile.ui.users;

import android.app.Application;
import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import java.util.LinkedList;
import java.util.List;

import ru.mrlargha.stemobile.data.Result;
import ru.mrlargha.stemobile.data.STERepository;
import ru.mrlargha.stemobile.data.model.SimpleServerReply;
import ru.mrlargha.stemobile.data.model.User;
import ru.mrlargha.stemobile.data.model.UsersReply;

public class UsersViewModel extends AndroidViewModel {

    private MutableLiveData<List<User>> usersLiveData = new MutableLiveData<>(new LinkedList<>());
    private MutableLiveData<Boolean> hasNetworkOperationInProgress = new MutableLiveData<>(false);
    private MutableLiveData<String> errorString = new MutableLiveData<>("");
    private STERepository mSTERepository;

    public UsersViewModel(@NonNull Application application) {
        super(application);
        mSTERepository = STERepository.getRepository(application);
        hasNetworkOperationInProgress.setValue(true);
        new GetUsersTask().execute();
    }

    void setUserPermission(User user) {
        hasNetworkOperationInProgress.setValue(true);
        new SetUserPermissionTask().execute(String.valueOf(user.getVk_id()), user.getPermissions());
    }

    MutableLiveData<Boolean> getHasNetworkOperationInProgress() {
        return hasNetworkOperationInProgress;
    }

    MutableLiveData<List<User>> getUsersLiveData() {
        return usersLiveData;
    }

    MutableLiveData<String> getErrorString() {
        return errorString;
    }

    private class GetUsersTask extends AsyncTask<Void, Void, List<User>> {

        @Override
        protected List<User> doInBackground(Void... voids) {
            Result<UsersReply> result = mSTERepository.getUsers();
            if (result instanceof Result.Success) {
                return ((Result.Success<UsersReply>) result).getData().getUsers();
            } else {
                errorString.postValue(((Result.Error) result).getErrorString());
                return new LinkedList<>();
            }
        }

        @Override
        protected void onPostExecute(List<User> users) {
            super.onPostExecute(users);
            if (!users.isEmpty()) {
                usersLiveData.setValue(users);
            }
            hasNetworkOperationInProgress.setValue(false);
        }
    }

    private class SetUserPermissionTask extends AsyncTask<String, Void, String> {
        @Override
        protected String doInBackground(String... strings) {
            Result<SimpleServerReply> result =
                    mSTERepository.setUserPermission(Integer.parseInt(strings[0]), strings[1]);
            if (result instanceof Result.Success) {
                return ((Result.Success<SimpleServerReply>) result).getData().getStatus();
            } else {
                return ((Result.Error) result).getErrorString();
            }
        }

        @Override
        protected void onPostExecute(String result) {
            super.onPostExecute(result);
            if (result.equals("ok")) {
                hasNetworkOperationInProgress.setValue(false);
            } else {
                errorString.postValue(result);
                new GetUsersTask().execute();
            }
        }
    }
}

File: app/src/main/java/ru/mrlargha/stemobile/ui/users/UsersViewModel.kt
package ru.mrlargha.stemobile.ui.users

import android.app.Application
import android.os.AsyncTask
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import ru.mrlargha.stemobile.data.Result
import ru.mrlargha.stemobile.data.STERepository
import ru.mrlargha.stemobile.data.STERepository.Companion.getRepository
import ru.mrlargha.stemobile.data.model.SimpleServerReply
import ru.mrlargha.stemobile.data.model.User
import ru.mrlargha.stemobile.data.model.UsersReply
import java.util.*

class UsersViewModel(application: Application) : AndroidViewModel(application) {
    val usersLiveData = MutableLiveData<List<User>>(LinkedList())
    val hasNetworkOperationInProgress = MutableLiveData(false)
    val errorString = MutableLiveData("")
    private val mSTERepository: STERepository?
    fun setUserPermission(user: User) {
        hasNetworkOperationInProgress.value = true
        SetUserPermissionTask().execute(user.vk_id.toString(), user.permissions)
    }

    private inner class GetUsersTask : AsyncTask<Void?, Void?, List<User>>() {
        override fun doInBackground(vararg params: Void?): List<User>? {
            val result: Result<*> = mSTERepository!!.users
            return if (result is Result.Success<*>) {
                (result as Result.Success<UsersReply?>).data!!.users
            } else {
                errorString.postValue((result as Result.Error).errorString)
                LinkedList()
            }
        }

        override fun onPostExecute(users: List<User>) {
            super.onPostExecute(users)
            if (!users.isEmpty()) {
                usersLiveData.value = users
            }
            hasNetworkOperationInProgress.value = false
        }
    }

    private inner class SetUserPermissionTask : AsyncTask<String?, Void?, String?>() {
        override fun doInBackground(vararg params: String?): String? {
            val result: Result<*> = mSTERepository!!.setUserPermission(params[0]!!.toInt(), params[1])
            return if (result is Result.Success<*>) {
                (result as Result.Success<SimpleServerReply?>).data!!.status
            } else {
                (result as Result.Error).errorString
            }
        }

        override fun onPostExecute(result: String?) {
            super.onPostExecute(result)
            if (result == "ok") {
                hasNetworkOperationInProgress.setValue(false)
            } else {
                errorString.postValue(result)
                GetUsersTask().execute()
            }
        }
    }

    init {
        mSTERepository = getRepository(application)
        hasNetworkOperationInProgress.value = true
        GetUsersTask().execute()
    }
}